frontend/src/metabase/hoc/UserHasSeen/UserHasSeenAllContext.tsx

```diff
@@ -0,0 +1,86 @@
+import {
+  type PropsWithChildren,
+  createContext,
+  useCallback,
+  useMemo,
+  useState,
+} from "react";
+
+import { useUserKeyValue } from "metabase/hooks/use-user-key-value";
+
+const EMPTY_ARRAY: string[] = [];
+
+export interface UserHasSeenAllContextProps {
+  upsertBadge: ({ key, value }: { key: string; value: boolean }) => void;
+  removeBadge: ({ key }: { key: string }) => void;
+  handleOpen: () => void;
+  hasSeenAll: boolean;
+}
+
+export const UserHasSeenAllContext =
+  createContext<UserHasSeenAllContextProps | null>(null);
+
+export const UserHasSeenAllProvider = ({
+  menuKey,
+  children,
+}: PropsWithChildren<{ menuKey: string }>) => {
+  const contextValue = useUserHasSeenAll(menuKey);
+
+  return (
+    <UserHasSeenAllContext.Provider value={contextValue}>
+      {children}
+    </UserHasSeenAllContext.Provider>
+  );
+};
+
+const useUserHasSeenAll = (menuKey: string) => {
+  const [badges, setBadges] = useState<[string, boolean][]>([]);
+
+  const upsertBadge = useCallback(
+    ({ value, key }: { value: boolean; key: string }) => {
+      setBadges(s => {
+        const badgeAlreadyInArray =
+          s.findIndex(([k, v]) => k === key && v === value) >= 0;
+        if (badgeAlreadyInArray) {
+          return s;
+        }
+
+        return [...s.filter(([k]) => k !== key), [key, value]];
+      });
+    },
+    [],
+  );
+  const removeBadge = useCallback(({ key }: { key: string }) => {
+    setBadges(s => [...s.filter(([k]) => k !== key)]);
+  }, []);
+
+  const { value: seenBadges, setValue: setSeenBadges } = useUserKeyValue({
+    namespace: "indicator-menu",
+    key: menuKey,
+    defaultValue: EMPTY_ARRAY,
+  });
+
+  const unseenBadges = useMemo(() => {
+    return badges
+      .filter(([key]) => !seenBadges.includes(key))
```

Comment:
would these lists ever get really large? this is O(n^2) but if n is small it's probs fine.

Reply:
They really shouldn't, but I can try and optimize a bit haha

Reply:
Don't sweat it if it won't get long. I wasn't sure if keys might include things like ids, and if this could get quite large in the future.